pddl/parser/domain.py

```diff
@@ -289,47 +292,24 @@ def constant(self, args):
 
     def atomic_formula_skeleton(self, args):
         """Process the 'atomic_formula_skeleton' rule."""
-        name = args[1]
+        predicate_name = args[1]
         variable_data: Dict[str, Set[str]] = args[2]
-        variables = [Variable(name, tags) for name, tags in variable_data.items()]
-        return Predicate(name, *variables)
-
-    def typed_list_name(self, args) -> Dict[str, Optional[str]]:
-        """
-        Process the 'typed_list_name' rule.
-
-        Return a dictionary with as keys the names and as value the type of the name.
-
-        Steps:
-        - if the '-' symbol is not present, then return the list of names
-        - if the '-' symbol is present, parse the names with their type tags
-
-        :param args: the argument of this grammar rule
-        :return: a typed list (name)
-        """
-        type_sep_index = safe_index(args, Symbols.TYPE_SEP.value)
-
-        if type_sep_index is None:
-            # simple list of names
-            return self._parse_simple_typed_list(args, check_for_duplicates=True)
-
-        # if we are here, the matched pattern is: [name_1, ..., name_n], "-", parent_name, other_typed_list_dict
-        # make sure there are only two tokens after "-"
-        assert_(len(args[type_sep_index:]) == 3, "unexpected parser state")
-
-        names: Tuple[str, ...] = tuple(args[:type_sep_index])
-        parent_name: str = str(args[type_sep_index + 1])
-        other_typed_list_dict: Mapping[str, Optional[str]] = args[type_sep_index + 2]
-        new_typed_list_dict: Mapping[str, Optional[str]] = {
-            obj: parent_name for obj in names
-        }
-
-        # check type conflicts
-        self._check_duplicates(other_typed_list_dict.keys(), new_typed_list_dict.keys())
-
-        return {**new_typed_list_dict, **other_typed_list_dict}
+        variables = [
+            Variable(var_name, tags) for var_name, tags in variable_data.items()
+        ]
+        return Predicate(predicate_name, *variables)
+
+    def typed_list_name(self, args) -> Dict[name, Optional[name]]:
+        """Process the 'typed_list_name' rule."""
+        try:
+            types_index = TypesIndex.parse_typed_list(args)
+            return types_index.get_typed_list_of_names()
+        except ValueError as e:
+            raise PDDLParsingError(
+                f"error while parsing tokens {list(map(str, args))}: {str(e)}"
+            ) from None
 
-    def typed_list_variable(self, args) -> Dict[str, Set[str]]:
+    def typed_list_variable(self, args) -> OrderedDictType[name, Set[name]]:
```

Comment:
`OrderedDict` is essential, predicates variables and action parameters require an ordered (possibly typed) list of variables.

The same cannot be said for `forall`/`exists` expressions, but having them ordered does not harm.